neonvm/controllers/vm_controller_unit_test.go

```diff
@@ -0,0 +1,262 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	vmv1 "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+)
+
+type mockRecorder struct {
+	mock.Mock
+}
+
+func (m *mockRecorder) Event(object runtime.Object, eventtype, reason, message string) {
+	m.Called(object, eventtype, reason, message)
+}
+
+func (m *mockRecorder) Eventf(object runtime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+	m.Called(object, eventtype, reason, messageFmt, args)
+}
+
+func (m *mockRecorder) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
+	m.Called(object, annotations, eventtype, reason, messageFmt, args)
+}
+
+// defaultVm returns a VM which is similar to what we can reasonably
+// expect from the control plane.
+func defaultVm() *vmv1.VirtualMachine {
+	return &vmv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-vm",
+			Namespace: "default",
+		},
+		Spec: vmv1.VirtualMachineSpec{
+			EnableSSH:          lo.ToPtr(false),
+			EnableAcceleration: lo.ToPtr(true),
+			//nolint:exhaustruct // This is a test
+			Guest: vmv1.Guest{
+				KernelImage:         lo.ToPtr("kernel-img"),
+				AppendKernelCmdline: nil,
+				CPUs: vmv1.CPUs{
+					Min: lo.ToPtr(vmv1.MilliCPU(1000)),
+					Max: lo.ToPtr(vmv1.MilliCPU(2000)),
+					Use: lo.ToPtr(vmv1.MilliCPU(1500)),
+				},
+				MemorySlots: vmv1.MemorySlots{
+					Min: lo.ToPtr(int32(512)),
+					Max: lo.ToPtr(int32(2048)),
+					Use: lo.ToPtr(int32(1024)),
+				},
+				MemorySlotSize: *resource.NewQuantity(1, resource.DecimalSI),
+			},
+		},
+		Status: vmv1.VirtualMachineStatus{
+			Phase:         "",
+			Conditions:    []metav1.Condition{},
+			RestartCount:  0,
+			PodName:       "",
+			PodIP:         "",
+			ExtraNetIP:    "",
+			ExtraNetMask:  "",
+			Node:          "",
+			CPUs:          lo.ToPtr(vmv1.MilliCPU(100)),
+			MemorySize:    resource.NewQuantity(123, resource.DecimalSI),
+			SSHSecretName: "",
+		},
```

Comment:
IIUC the VM starts with a completely empty status object and it's up to the controller to fill it out after reconciling. Should we have all fields empty here? (instead of including cpus & memory size?)